src/Lecture2.hs

```diff
@@ -1,170 +1,277 @@
-course but you can if you like challenges :)
--}
-
+-- |
+-- Module                  : Lecture2
+-- Copyright               : (c) 2021-2022 Haskell Beginners 2022 Course
+-- SPDX-License-Identifier : MPL-2.0
+-- Maintainer              : Haskell Beginners 2022 Course <haskell.beginners2022@gmail.com>
+-- Stability               : Stable
+-- Portability             : Portable
+--
+-- Exercises for the Lecture 2 of the Haskell Beginners course.
+--
+-- As in the previous section, implement functions and provide type
+-- signatures. If the type signature is not already written, write the
+-- most polymorphic type signature you can.
+--
+-- Unlike exercises to Lecture 1, this module also contains more
+-- challenging exercises. You don't need to solve them to finish the
+-- course but you can if you like challenges :)
 module Lecture2
-    ( -- * Normal
-      lazyProduct
-    , duplicate
-    , removeAt
-    , evenLists
-    , dropSpaces
-
-    , Knight (..)
-    , dragonFight
-
-      -- * Hard
-    , isIncreasing
-    , merge
-    , mergeSort
-
-    , Expr (..)
-    , Variables
-    , EvalError (..)
-    , eval
-    , constantFolding
-    ) where
+  ( -- * Normal
+    lazyProduct,
+    duplicate,
+    removeAt,
+    evenLists,
+    dropSpaces,
+    Knight (..),
+    Dragon (..),
+    DragonType (..),
+    Chest (..),
+    Attack (..),
+    Health (..),
+    Endurance (..),
+    Experience (..),
+    Gold (..),
+    FightResult (..),
+    Treasure (..),
+    Reward (..),
+    dragonFight,
+
+    -- * Hard
+    isIncreasing,
+    merge,
+    mergeSort,
+    Expr (..),
+    Variables,
+    EvalError (..),
+    eval,
+    constantFolding,
+  )
+where
 
 -- VVV If you need to import libraries, do it after this line ... VVV
+import Data.Char (isSpace)
+import Data.Maybe ()
 
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
-
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (x : xs) = x * lazyProduct xs
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate = concatMap (replicate 2)
```

Comment:
That's a correct solution 👍🏻 However, the existing implementation allocates temporary lists inside and appends them with `++`.

It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible.